### PR TITLE
*: add query metric http api

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
* Add query metric http api. `/pd/api/v1/metric/query`.
* Add `metric-query-address` config item to indicate the metric query address.

This is for https://github.com/pingcap/tidb/issues/13567 project.

### What is changed and how it works?
Currently, PD does not support query metric by itself, so just redirect the metric query to `metric-query-address`. After PD support storage metric data, PD can handle the query request by itself.

`metric-query-address` should be dynamic adjustment. Because the user can choose which metric data should be read. So I add this config item in `pd-server` for dynamic adjustment.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has configuration change

Side effects
Related changes

### CODE REVIEW

File: conf/config.toml [hunk 1]
```diff
@@ -54,6 +54,12 @@ interval = "15s"
 ## prometheus pushgateway address, leaves it empty will disable prometheus.
 address = ""
 
+[pd-server]
+## the metric storage is the cluster metric storage. This is use for query metric data.
+## Currently we use prometheus as metric storage, we may use PD/TiKV as metric storage later.
+## For usability, recommended to temporarily set it to the prometheus address, eg: http://127.0.0.1:9090
+metric-storage = ""
```
Comment (rleungx) [inline conf/config.toml]: It sounds like a kind of storage, but actually it is an address.

File: server/api/router.go [hunk 2]
```diff
@@ -144,6 +144,9 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	rootRouter.Handle("/api/v1/health", newHealthHandler(svr, rd)).Methods("GET")
 	rootRouter.Handle("/api/v1/diagnose", newDiagnoseHandler(svr, rd)).Methods("GET")
 	rootRouter.HandleFunc("/api/v1/ping", func(w http.ResponseWriter, r *http.Request) {}).Methods("GET")
+	// metric query use to query metric data, the protocol is compatible with prometheus.
+	rootRouter.Handle("/api/v1/metric/query", newQueryMetric(svr)).Methods("GET", "POST")
+	rootRouter.Handle("/api/v1/metric/query_range", newQueryMetric(svr)).Methods("GET", "POST")
```
Comment (rleungx) [inline server/api/router.go]: What is the difference between line 148 and line 149?

Comment (crazycs520) [reply server/api/router.go]: The URI path `query ` is difference with `query_range `.

File: conf/config.toml (hunk 1 above)
Comment (crazycs520) [reply conf/config.toml]: Ye, I already talk with @lonng , @siddontang, they prefer this. :)

File: server/api/router.go (hunk 2 above)
Comment (rleungx) [reply server/api/router.go]: Oh. I just realized this is used by Prometheus HTTP API.

**Comment by disksing** [comment]
Hi @crazycs520 , you need to update API docs too. See https://github.com/pingcap/pd/blob/master/docs/development.md#changing-apis

File: server/api/router.go [hunk 3]
```diff
@@ -144,6 +144,9 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	rootRouter.Handle("/api/v1/health", newHealthHandler(svr, rd)).Methods("GET")
 	rootRouter.Handle("/api/v1/diagnose", newDiagnoseHandler(svr, rd)).Methods("GET")
 	rootRouter.HandleFunc("/api/v1/ping", func(w http.ResponseWriter, r *http.Request) {}).Methods("GET")
+	// metric query use to query metric data, the protocol is compatible with prometheus.
+	rootRouter.Handle("/api/v1/metric/query", newQueryMetric(svr)).Methods("GET", "POST")
```
Comment (nolouch) [inline server/api/router.go]: need to update the API documentation in `api.raml`.

File: conf/config.toml (hunk 1 above)
Comment (lhy1024) [reply conf/config.toml]: how about `metric-storage-url`?

File: server/api/router.go (hunk 2 above)
Comment (lhy1024) [inline server/api/router.go]: It seem to be more restful to use `query-range` or `range`?

File: server/api/metric.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	switch u.Scheme {
+	case "http", "https":
+		// Replace the pd path with the prometheus http API path.
+		r.URL.Path = strings.Replace(r.URL.Path, "pd/api/v1/metric", "api/v1", 1)
+		newCustomReverseProxies([]url.URL{*u}).ServeHTTP(w, r)
+	default:
+		// TODO: Support read data by self after support store metric data in PD/TiKV.
+		http.Error(w, fmt.Sprintf("schema of metric storage address is no support, address: %v", metricAddr), http.StatusInternalServerError)
```
Comment (lhy1024) [inline server/api/metric.go]:
```suggestion
		http.Error(w, fmt.Sprintf("schema of metric storage address is no supported, address: %v", metricAddr), http.StatusInternalServerError)
```

File: server/api/router.go (hunk 2 above)
Comment (crazycs520) [reply server/api/router.go]: @lhy1024  `query_range` is compatible with Prometheus api. https://prometheus.io/docs/prometheus/latest/querying/api/

File: conf/config.toml (hunk 1 above)
Comment (crazycs520) [reply conf/config.toml]: The usage of `metric-storage` will not only indicate the `url address`.

**Pull request merged**